### ISSUE TITLE
Msr Exit doesn't work?

### ISSUE THREAD

**Issue by yourapple** [tandasat/HyperPlatform#50]
I compiled the default project and tested it on VMware win7 (x64) Sp1.

VMware has been configured according to the instructions.

Hypervisor.cpuid.v0 = "FALSE"

Mce.enable = "TRUE"

The following is the output information of the WinDbg. There is no trigger information for MSR exit.

I also added print information in VmmpHandleMsrAccess, and I didn't have any output.

19:58:46.917	INF	#3	    4	   52	System         	Initialized successfully.
19:58:46.949	INF	#3	    4	   52	System         	The VMM has been installed.
20:04:32.226	INF	#1	    4	   68	System         	Uninstalling VMM.
20:04:32.242	INF	#0	    4	   68	System         	Terminating VMX for the processor 0.
20:04:32.273	DBG	#0	    4	   68	System         	Context at FFFFF880047CBA20 FFFFFA8003C1F070
20:04:32.273	DBG	#0	    4	   68	System         	Used pre-allocated entries  =  3 / 50
20:04:32.304	INF	#1	    4	   68	System         	Terminating VMX for the processor 1.
20:04:32.320	DBG	#1	    4	   68	System         	Context at FFFFF880047CBA20 FFFFFA80039CD7B0
20:04:32.320	DBG	#1	    4	   68	System         	Used pre-allocated entries  =  2 / 50
20:04:32.366	INF	#2	    4	   68	System         	Terminating VMX for the processor 2.
20:04:32.398	DBG	#2	    4	   68	System         	Context at FFFFF880047CBA20 FFFFFA8003AAB330
20:04:32.398	DBG	#2	    4	   68	System         	Used pre-allocated entries  =  2 / 50
20:04:32.429	INF	#3	    4	   68	System         	Terminating VMX for the processor 3.
20:04:32.444	DBG	#3	    4	   68	System         	Context at FFFFF880047CBA20 FFFFFA800196F170
20:04:32.444	DBG	#3	    4	   68	System         	Used pre-allocated entries  =  2 / 50
20:04:32.507	DBG	#3	    4	   68	System         	Freeing shared data...
20:04:32.522	INF	#3	    4	   68	System         	The VMM has been uninstalled.
20:04:32.538	INF	#1	    4	   68	System         	FunctionName(Line)                           ,Execution Count     ,Elapsed Time        
20:04:32.554	INF	#1	    4	   68	System         	VmmpHandleCpuid(438)                         ,                  80,                 516,
20:04:32.569	INF	#1	    4	   68	System         	VmmpHandleVmExit(228)                        ,               50554,              749860,
20:04:32.585	INF	#1	    4	   68	System         	EptHandleEptViolation(736)                   ,                  34,                2168,
20:04:32.616	INF	#1	    4	   68	System         	VmmpHandleEptViolation(1296)                 ,                  34,                2365,
20:04:32.647	INF	#3	    4	   68	System         	VmmpHandleCrAccess(1123)                     ,               46825,              255310,
20:04:32.663	INF	#1	    4	   68	System         	VmmpHandleCrAccess(1088)                     ,               46727,              408260,
20:04:32.678	INF	#1	    4	   68	System         	VmmpHandleGdtrOrIdtrAccess(648)              ,                 666,               50877,
20:04:32.756	INF	#1	    4	   68	System         	VmmpHandleLdtrOrTrAccess(738)                ,                 661,               12659,
20:04:32.772	INF	#1	    4	   68	System         	VmmpHandleDrAccess(834)                      ,                2309,               15240,
20:04:32.803	DBG	#1	    4	   68	System         	Finalizing... (Max log usage = 445/65536 bytes)
20:04:32.819	INF	#1	    4	   68	System         	Bye!

Is it wrong with me? Does it need additional configuration to trigger the MSR?

**Comment by tandasat** [comment]
It seems MSR access does not happen very often on Windows 7. Try generating more activities; I used CPU-Z and got few access to `VmmpHandleMsrReadAccess`. No modification in code should be needed.

```
09:37:32.118 INF #0     4    64 System          FunctionName(Line)                           ,Execution Count     ,Elapsed Time        
09:37:32.118 INF #0     4    64 System          VmmpHandleCpuid(407)                         ,                 279,                1263,
09:37:32.118 INF #0     4    64 System          VmmpHandleVmExit(224)                        ,               46239,              840087,
09:37:32.118 INF #0     4    64 System          EptHandleEptViolation(672)                   ,                4475,              100924,
09:37:32.118 INF #0     4    64 System          VmmpHandleEptViolation(1078)                 ,                4475,              121052,
09:37:32.118 INF #0     4    64 System          VmmpHandleCrAccess(921)                      ,               40506,              352368,
09:37:32.118 INF #0     4    64 System          VmmpHandleCrAccess(893)                      ,               40534,              625683,
09:37:32.118 INF #0     4    64 System          VmmpHandleMsrReadAccess(473)                 ,                 925,                3786,
09:37:32.118 INF #0     4    64 System          VmmpHandleGdtrOrIdtrAccess(559)              ,                  12,                 131,
09:37:32.118 INF #0     4    64 System          VmmpHandleDrAccess(726)                      ,                   5,                  29,
````

**Comment by yourapple** [comment]
Yes, I tested CPU-Z for MSR exit.



I used to test it with Securable before I thought Securable would read MSR information, and it didn't seem to be.



But I found that the default project will trigger PatchGuard on win7 (x64) SP1, and the following is part of the blue screen. It seems that the information of GDT is detected.

The test way is to only load the drive, and then wait a moment to trigger.

CRITICAL_STRUCTURE_CORRUPTION (109)
This bugcheck is generated when the kernel detects that critical kernel code or
data have been corrupted. There are generally three causes for a corruption:
1) A driver has inadvertently or deliberately modified critical kernel code
 or data. See http://www.microsoft.com/whdc/driver/kernel/64bitPatching.mspx
2) A developer attempted to set a normal kernel breakpoint using a kernel
 debugger that was not attached when the system was booted. Normal breakpoints,
 "bp", can only be set if the debugger is attached at boot time. Hardware
 breakpoints, "ba", can be set at any time.
3) A hardware corruption occurred, e.g. failing RAM holding kernel code or data.
Arguments:
Arg1: a3a039d896da8849, Reserved
Arg2: b3b7465ee958c4b3, Reserved
Arg3: fffff80000b95000, Failure type dependent information
Arg4: 0000000000000003, Type of corrupted region, can be
	0 : A generic data region
	1 : Modification of a function or .pdata
	2 : A processor IDT
	3 : A processor GDT
	4 : Type 1 process list corruption
	5 : Type 2 process list corruption
	6 : Debug routine modification
	7 : Critical MSR modification

Debugging Details:
------------------


BUGCHECK_STR:  0x109

DEFAULT_BUCKET_ID:  VISTA_DRIVER_FAULT

PROCESS_NAME:  System

CURRENT_IRQL:  2

LAST_CONTROL_TRANSFER:  from fffff80003fcdd92 to fffff80003ede490

SYMBOL_ON_RAW_STACK:  1

STACK_ADDR_RAW_STACK_SYMBOL: fffff880047bcfc0

STACK_COMMAND:  dds 047BCFC0-0x20 ; kb
